### PR TITLE
Hook spellcheck flow up to CKEditor events

### PR DESCRIPTION
**Pull request by caffodian**
Events will let us better write tests that test the spellcheck end to end.  For example we'll be able to mock HTTP properly to have certain words be typos, or cause certain steps to fail, make sure certain words were passed along, etc.

### CODE REVIEW

File: plugins/nanospell/plugin.js [hunk 1]
```diff
@@ -341,23 +348,36 @@
 				clearAllSpellCheckingSpans(editor.editable());
 			}
 
-			function checkNow() {
+			function checkNow(root) {
```
Comment (caffodian) [inline plugins/nanospell/plugin.js:353]: "root" param is mostly unused for now.  the idea is that the emitted events would provide a root.  but we don't currently use it for anything.

File: plugins/nanospell/plugin.js [hunk 2]
```diff
@@ -341,23 +348,36 @@
...
-						send(words);
-					}
+					editor.fire(EVENT_NAMES.START_SCAN_WORDS);
+				}
+			}
+
+			function scanWords(event) {
+				var words;
+				if (event.data) {
+					// TODO this would be the case where an element is passed in, we don't handle it
```
Comment (caffodian) [inline plugins/nanospell/plugin.js:370]: next issue. related to the root thing above.

File: plugins/nanospell/plugin.js [hunk 3]
```diff
@@ -341,23 +348,36 @@
...
 				if (commandIsActive) {
 
 					self._spellCheckInProgress = true;
 
-					var words = getWords(editor.document.$.body, maxRequest);
-					if (words.length == 0) {
-						render();
-					} else {
-						send(words);
-					}
```
Comment (caffodian) [inline plugins/nanospell/plugin.js:357]: moved down and replaced with events

File: plugins/nanospell/plugin.js [hunk 4]
```diff
@@ -455,25 +478,25 @@
...
 				var bookmarks = editor.getSelection().createBookmarks();
 				clearAllSpellCheckingSpans(editor.editable());
 				self.markAllTypos(editor);
 				editor.getSelection().selectBookmarks(bookmarks);
 
-				editor.fire('SpellcheckStart');
-				editor.nanospellstarted = true;
 				self._spellCheckInProgress = false;
 				self._timer = null;
+				editor.fire(EVENT_NAMES.SPELLCHECK_COMPLETE);
```
Comment (caffodian) [inline plugins/nanospell/plugin.js:497]: this event is mostly for testing

File: plugins/nanospell/plugin.js [hunk 5]
```diff
@@ -520,19 +543,8 @@
...
-					fullTextContext = '';
-
-				range.selectNodeContents(editor.editable());
-
-				var iterator = range.createIterator();
-				while (( block = iterator.getNextParagraph() )) {
-					fullTextContext += block.getText() + ' ';
-				}
-
-				var matches = fullTextContext.match(wordTokenizer());
```
Comment (caffodian) [inline plugins/nanospell/plugin.js:535]: this method was split into multiples.  and tried to split it into phases

File: plugins/nanospell/plugin.js [hunk 6]
```diff
@@ -70,11 +77,11 @@
...
 				( !node.isReadOnly() ) &&   // or read only
 				isNotBookmark(node) && // and isn't a fake bookmarking node
-				// tables and list items can get a bit weird with getNextParagraph()
-				// for example causing list item descendants to be included as part of the original list item
-				// and also individually as their own paragraph-like elements
-				(path.blockLimit ? path.blockLimit.equals(startNode): true) && // check we don't enter another block-like element
-				(path.block ? path.block.equals(startNode): true); // check we don't enter nested blocks (special list case since it's not considered a limit)
+					// tables and list items can get a bit weird with getNextParagraph()
+					// for example causing list item descendants to be included as part of the original list item
+					// and also individually as their own paragraph-like elements
```
Comment (bocribbz) [inline plugins/nanospell/plugin.js]: what with the xtra comments indent? makes it more visible?

Comment (caffodian) [inline plugins/nanospell/plugin.js]: I think the IDE code formatter thinks this is a continuation of the previous comment.  this is not entirely correct.  will fix or probably move the details of the overly complicated case to below.

File: plugins/nanospell/plugin.js [hunk 7]
```diff
@@ -543,14 +558,39 @@
 					if (!uniqueWords[word] && self.validWordToken(word) && (typeof(spellcache[word]) === 'undefined')) {
 						words.push(word);
 						uniqueWords[word] = true;
-						if (words.length >= max) {
-							return words;
-						}
+						//if (words.length >= max) {
+						//	return words;
+						//}
```
Comment (bocribbz) [inline plugins/nanospell/plugin.js]: is it useful to keep the commented lines here?

Comment (caffodian) [inline plugins/nanospell/plugin.js]: Hmm. Probably not. It's a bit of a reminder that I removed the max words per request which wasn't really working, but I suspect I won't need to bring it back.

**Pull request merged**